sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala

```diff
@@ -204,6 +204,84 @@ class SubquerySuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  test("SPARK-29145: JOIN Condition use QueryList") {
+    withTempView("s1", "s2", "s3") {
+      Seq(1, 3, 5, 7, 9).toDF("id").createOrReplaceTempView("s1")
+      Seq(1, 3, 4, 6, 9).toDF("id").createOrReplaceTempView("s2")
+      Seq(3, 4, 6, 9).toDF("id").createOrReplaceTempView("s3")
+
+      checkAnswer(
+        sql("SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id IN (select 9)"),
+        Row(9) :: Nil)
+
+      checkAnswer(
+        sql("SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id NOT IN (select 9)"),
+        Row(1) :: Row(3) :: Nil)
+
+      // case `IN`
+      checkAnswer(
+        sql("SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id IN (select id from s3)"),
+        Row(3) :: Row(9) :: Nil)
+
+      checkAnswer(
+        sql("SELECT s1.id as id2 from s1 LEFT SEMI JOIN s2 " +
```

Comment:
nit: can you follow the format of the other tests? In multi-line cases, the format seems to be like this;
```
      sql("""
            |
            | ...
            ...
            |  ) """.stripMargin)
```

Reply:
Changed.